Add delegate.cash support

datadanne opened the pull request:
Updates the contract to allow a delegated wallet to train/pilot/park rigs using delegate.cash.
Adds a "Act as" button to the garage top header that is only shown if the currently connected wallet is registered as a delegate for a vault for the Rigs contract. A click on the button lets the user change wallet that they want to make actions for:

![image](https://user-images.githubusercontent.com/656107/229504113-c0a52873-28aa-4269-9889-0539444518d7.png)
![image](https://user-images.githubusercontent.com/656107/229504127-b88657a9-95f6-4a73-a9c1-2b4b1074f3e4.png)


Closes RIG-17

sanderpick commented on ethereum/contracts/TablelandRigs.sol:
```diff
@@ -446,11 +464,25 @@ contract TablelandRigs is
-        if (ownerOf(tokenId) != _msgSenderERC721A())
-            revert ITablelandRigPilots.Unauthorized();
+        address sender = _msgSenderERC721A();
+        address tokenOwner = ownerOf(tokenId);
+
+        if (
+            _delegateCash.checkDelegateForToken(
+                sender,
+                tokenOwner,
+                address(this),
+                tokenId
+            )
+        ) {
+            sender = tokenOwner;
+        }
 
-        _pilots.trainRig(_msgSenderERC721A(), tokenId);
+        if (tokenOwner != sender) revert ITablelandRigPilots.Unauthorized();
```
this could be a function/modifier to save on contract size since it's reused three times

sanderpick commented on ethereum/contracts/DelegateCashMock.sol:
```diff
@@ -0,0 +1,40 @@
+    ) external;
+}
+
+contract DelegateCashMock is MockDelegateCashInterface {
```
so is there one "real" public delegatecash contract out in the wild that we'll use?

datadanne replied:
Yeah. delegate.cash is deployed to the same address on most networks (https://docs.delegate.cash/delegatecash/technical-documentation/delegation-registry/contract-addresses), but we can't add new constants to the contract since it is upgradeable so I think we need to use a setter instead?

sanderpick replied:
we could add new consts... I think it's just a matter of if we want to. may as well to avoid a new method? note: consts can change with upgrades since they don't use contract storage.

datadanne replied:
yeah I agree. I was going to do that first but for some reason I thought that we couldn't add a constant to an upgradeable contract. fixed!

datadanne commented on ethereum/contracts/interfaces/IDelegationRegistry.sol:
The full implementation is available here https://docs.delegate.cash/delegatecash/technical-documentation/delegation-registry/idelegationregistry.sol, but since we only use one method it felt simpler for testing purposes to use a reduced version. Do you have a preference or do you know if there are any best practices @sanderpick?

sanderpick replied:
What you did (the minimal interface) looks! I've seen that pattern elsewhere... not sure about solidity best practice there but in general software eng, declaring the subset of an interface that you're using close to the impl is a healthy practice.

sanderpick commented:
@datadanne it would be good to get this PR resolved this week!

datadanne commented:
> @datadanne it would be good to get this PR resolved this week!

Agreed! I think I addressed all your comments last week, and I just rebased it so it should be ready for review again.

sanderpick commented:
I realized I should have merged this on Friday since you were out. We're gonna try merging normal work into a staging branch to avoid dangling PRs. I'll create one on this repo. We can merge to there and ship it next week. 

Bug fixe PRs can just be merged to main.

datadanne commented:
> I realized I should have merged this on Friday since you were out. We're gonna try merging normal work into a staging branch to avoid dangling PRs. I'll create one on this repo. We can merge to there and ship it next week.
> 
> Bug fixe PRs can just be merged to main.

Sounds good! And it kind of worked out well that none of us merged it last week since it would have deployed the UI change right away, but the UI won't work until we've deployed the contract update 👍
